funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/result_queue_publisher.py

```diff
@@ -0,0 +1,82 @@
+import logging
+
+import pika
+
+logger = logging.getLogger(__name__)
+
+
+class ResultQueuePublisher:
+    """ResultPublisher publishes results to a topic exchange, with the {endpoint_id}.results
+    as a routing key.
+    """
+
+    def __init__(
+        self,
+        endpoint_id: str,
+        pika_conn_params: pika.connection.Parameters,
+        exchange="results",
+    ):
+        """
+        Parameters
+        ----------
+        endpoint_uuid: str
+            Endpoint UUID string used to identify the endpoint
+        pika_conn_params: pika.connection.Parameters
+            Pika connection parameters to connect to RabbitMQ
+        exchange: str
+            Exchange name. Default: "result"
+        """
+        self.endpoint_id = endpoint_id
+        self.routing_key = f"{self.endpoint_id}.results"
+        self.params = pika_conn_params
+        self.params.heartbeat = 30
+        self.params.blocked_connection_timeout = 60
+        self.exchange = exchange
+        self.exchange_type = "topic"
+        self._channel = None
+        self._connection = None
```

Comment:
It's just convention to have `connect` be a separate step. Not sure if this makes much of a difference. I think it's cleaner to have close and connect and be explicit. 
```
# create steps
pub = publisher()
pub.connect()

# Reconnect steps
pub.close()
pub.connect() 
```

Reply:
Yeah, that's reasonable. I'm not sure it makes any difference for this. I was just looking at this and thinking about the fact that the type of `self._connection` is `Connection | None`, rather than `Connection`.

I _think_ I would like to see the connection wrapped in a property which raises an error if it's read while the value is `None`.
e.g.
```
@property
def connection(self) -> Connection:
    if self._connection is None:
        raise ValueError("connection was not set")
    return self._connection 
```

The reason is that then we're handling something of type `Connection`, not `Connection | None`. So we have fewer kinds of incorrect usage that will bomb out in unexpected ways.

Relatedly, I would also like to see these classes implementing the context manager interface, which ensures that `__exit__` runs and we get clean shutdown. These are just some basic steps we can take when setting this up to ensure correctness.

Reply:
This makes sense, but I'm not sure if we want to make these changes in this round. I suspect we'll need to move these rabbitMQ bits into `funcx-common` and at that point, I'd like to reorganize the subscribers and publishers inherit from a parent class.

Reply:
I'd like to see these changes but, unlike some of the other comments, there's nothing blocking here. I've marked this resolved.